### PR TITLE
Add support for platform libraries

### PR DESCRIPTION
**Pull request by edsko**
This PR adds support for building platform libraries. A stanza for a platform library looks something like

```
platform-library test-package
  type:                native-shared

  if os(Windows)
    options: standalone
    mod-def-file: TestPackage.def

  other-modules:       MyPlatformLib.Hello
                       MyPlatformLib.SomeBindings
  build-depends:       base >=4.7 && <4.9
  hs-source-dirs:      src
  c-sources:           csrc/MyPlatformLibWrapper.c
  default-language:    Haskell2010
```

where `native-shared` means that we want to build a native shared library (`.so` on Linux, `.dylib` on OSX, `.dll` on Windows). The parser also recognizes `native-static` but this is not currently supported anywhere. The `standalone` option means that the we merge all library dependencies into the dynamic library (i.e., ghc options `-shared -static`), rather than make the created dynamic library just record its dependencies (ghc options `-shared -dynamic`); it is currently compulsory on Windows and unsupported anywhere else. The `mod-def-file` can be used to specify a [module definition file](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/win32-dlls.html), and is also Windows specific.

This probably needs a fairly careful review, in particular in the definition of `gbuild`, which is the old `buildOrReplExe` and now deals with both executables and platform libraries.

Note: there will be a minor clash between this and the independent goals stuff (in particular tracking fine grained dependencies, #2514).

/cc @dcoutts

### CODE REVIEW

**Comment by 23Skidoo** [comment]
Needs documentation. Will take a closer look soon.

**Comment by dcoutts** [comment]
Ah yes, user guide. @edsko  and I will do that.

**Comment by 23Skidoo** [comment]
Also would be nice to have a test. And there's a Travis failure.

File: Cabal/Distribution/Simple/Haddock.hs [hunk 1]
```diff
@@ -224,6 +224,7 @@ haddock pkg_descr lbi suffixes flags = do
                          version
               let libArgs' = commonArgs `mappend` libArgs
               runHaddock verbosity tmpFileOpts comp confHaddock libArgs'
+        CPLib  _ -> putStrLn "<<haddock CPLib>>"
```
Comment (23Skidoo) [inline Cabal/Distribution/Simple/Haddock.hs]: Looks suspicious.

Comment (edsko) [inline Cabal/Distribution/Simple/Haddock.hs]: Oh, whoops. Sorry about that. Will fix that (along with your other comments). Not yet sure why Travis is failing though.

Comment (23Skidoo) [inline Cabal/Distribution/Simple/Haddock.hs]: > Not yet sure why Travis is failing though.

Looks like a warning (we use `-Werror` on Travis).

**Comment by 23Skidoo** [comment]
Some general comments:

1. As a user, I find the notion of "platform library" as a first-class concept a bit confusing. While it is easy for a beginner to grasp what the other top-level entities (`library`, `executable`, `test`, `bench` etc.) in a .cabal file are, "platform library" is a non-standard Cabal-specific concept. The name is also a bit vague - without reading documentation, I couldn't have guessed that this refers to a DLL intended to be called from programs in other languages. Maybe "embedded library" would be better, though that could lead people to believe that it has something to do with embedded programming. Is there some accepted terminology for this concept? How do other languages/build tools implement it?
2. Platform libraries appear to be very similar to "normal" libraries. Have you considered making this functionality part of the existing `library` stanza? That is, something like:
```
library foo
    platform-library: True -- or embedded-library, or whatever
    if os(windows)
        standalone: True
        mod-def-file: path/to/file.def
    [...]
```
If the problem is that `library` must be unique in a package, maybe we should lift that restriction.

/cc @tibbe -- you might want to comment on this.

**Comment by dcoutts** [comment]
I'm happy to consider other names that make it clearer. Suggestions? "system-library", "native-library", "foreign-library"?

As for whether it should be merged with normal libraries, no I think it should be a separate stanza, because it's building a qualitative different artefact. These system libs are inherently less portable, sometimes completely platform specific.

About other libraries, yes I think we should have "private" Haskell libraries within a package, but I would keep those as portable Haskell libs.

**Comment by edsko** [comment]
(Also, compared to "normal" libraries, other library specific flags such as `exposed-modules` do not apply to platform/system/native libraries.)

**Comment by dcoutts** [comment]
These libs also cannot be listed in build-depends in other packages (though in principle could be listed as `extra-libraries:`, like we depend on other C libs).

This new stanza is intended to be extensible with more `type`s, for build artefacts like Windows assemblies, or OSX frameworks, or other platform-specific build artefacts that use platform-native foreign-code APIs.

**Comment by 23Skidoo** [comment]
Re: bikeshedding

> Suggestions? "system-library",  "native-library", "foreign-library"?
None of this sounds quite right. "System library" feels like it refers to something like `kernel32.dll`,
"native" usually refers to native code, "foreign" in Haskell context means non-Haskell code.

**Comment by edsko** [comment]
Of course, "foreign" doesn't necessary refer to non-Haskell code, we have both foreign imports and foreign exports, and a "platform-library" is typically one that has foreign exports.

**Comment by dcoutts** [comment]
@23Skidoo I'm no claiming foreign is perfect, but the rationale is that it's for consumption by foreign code, the API it presents is an FFI one.

**Comment by 23Skidoo** [comment]
Yes, "foreign" is probably the least bad sugggestion.

> Of course, "foreign" doesn't necessary refer to non-Haskell code, we have both foreign imports and foreign exports.

True, but usually unqualified "foreign" means non-Haskell, as in "foreign pointer".

**Comment by edsko** [comment]
| True, but usually unqualified "foreign" means non-Haskell, as in "foreign pointer".

Such a Haskell-centric view of the world.. oh wait :-P

**Comment by dcoutts** [comment]
I'm happy with "foreign-library"

**Comment by 23Skidoo** [comment]
I sort of like "foreign export library", but it's a bit long. I looked into how other languages call this, but there doesn't seem to be accepted terminology.

**Comment by edsko** [comment]
I realized that the naming of the generated library isn't quite right, and we are also missing an implementation of `copy`; fixing both of those.

**Comment by edsko** [comment]
Started to look at implementing a test, but had some difficulty getting the existing package tests to work on my system. I've made some changes to the test driver so that they now work; they should also make it easier to modify these tests further later should we need to.

However, with 7.4.2 I'm still getting one test error (with or without my changes):

```
  MultipleSource
    finds second source of multiple source:                      FAIL (1.80s)
      expected: 'cabal install' should succeed
        output: "/Users/e/wt/projects/cabal/cabal-install/dist/7.4.2/build/cabal/cabal --config-file=/Users/e/wt/projects/cabal/cabal-install/tests/PackageTests/cabal-config install q" in PackageTests/MultipleSource
      Resolving dependencies...
      Notice: installing into a sandbox located at
      /Users/e/wt/projects/cabal/cabal-install/tests/PackageTests/MultipleSource/.cabal-sandbox
      Configuring q-0.1.0.0...
      Building q-0.1.0.0...
      Failed to install q-0.1.0.0
      Build log ( /Users/e/wt/projects/cabal/cabal-install/tests/PackageTests/MultipleSource/.cabal-sandbox/logs/q-0.1.0.0.log ):
      Configuring q-0.1.0.0...
      Building q-0.1.0.0...
      setup-Simple-Cabal-1.14.0-x86_64-osx-ghc-7.4.2:
      dist/dist-sandbox-3b4eb8e0/package.conf.inplace: inappropriate type
      cabal: Error: some packages failed to install:
      q-0.1.0.0 failed during the building phase. The exception was:
      ExitFailure 1
```

I don't know what that's about.

**Comment by dcoutts** [comment]
@23Skidoo happy enough with `foreign-library`? If so we'll change the code (and rename things in the code to use the same terminology).

**Comment by 23Skidoo** [comment]
@dcoutts What do you think about `foreign-export-library`? IMO it's a bit clearer, if verbose.

**Comment by dcoutts** [comment]
Quick poll on `#hackage` didn't come up with a clear winner.

**Comment by simonmichael** [comment]
foreign-library was also my first guess, based on other usage of "foreign" in Haskell. One very slight issue, it supports the perception that Haskellers view the rest of the world with suspicion. Better if it could communicate "Haskell plays nicely with everything".

c-library, because library can be used by Haskell code, while this thing can be used by things which know the C ABI. Or portable-library, because C ABI is too limiting.

platform-library is not bad, makes sense to folks who know software dev jargon. It might be read as "the platform we are building on" which might be wrong.

native-library seems similar to platform-library, but clearer. It feels more capable and powerful than 
foreign-library. I don't quite get the objection to it above.

**Comment by edsko** [comment]
@23Skidoo , @dcoutts : Ok, added a test which builds and installs a platform library, and then attempts to link a C program against it. Verified that the test works on all supported platforms (Linux, OSX, Windows).

**Comment by 23Skidoo** [comment]
Your test seems to fail with GHC < 7.10 on Linux.

**Comment by edsko** [comment]
Darn. Looking at this now. Seems some library paths are different. Bear with me.

**Comment by edsko** [comment]
Ugh, what a @£@% mess. I think it should now work on Linux for all versions tested on Travis. Here's what was going wrong (I've also added this as a note in the source code itself):

------------------------

Suppose that the dynamic library depends on `base`, but not (directly) on `integer-gmp` (which, however, is a dependency of `base`). We will link the library as

    gcc ... -lHSbase-4.7.0.2-ghc7.8.4 -lHSinteger-gmp-0.5.1.0-ghc7.8.4 ...

However, on systems (like Ubuntu) where the linker gets called with `-as-needed` by default, the linker will notice that `integer-gmp` isn't actually a direct dependency and hence omit the link. 

Then when we attempt to link a C program against this dynamic library, the _static_ linker will attempt to verify that all symbols can be resolved.  The dynamic library itself does not require any symbols from `integer-gmp`, but `base` does. In order to verify that the symbols used by `base` can be resolved, the static linker needs to be able to _find_ integer-gmp. 

Finding the `base` dependency is simple, because the dynamic elf header (`readelf -d`) for the library that we have created looks something like

    (NEEDED) Shared library: [libHSbase-4.7.0.2-ghc7.8.4.so]
    (RPATH)  Library rpath: [/path/to/base-4.7.0.2:...]

This specifies the location of `integer-gmp` _in terms of_ the location of base (using the `$ORIGIN`) variable. But here's the crux: when the static linker attempts to verify that all symbols can be resolved, [**IT DOES NOT RESOLVE `$ORIGIN`**](http://stackoverflow.com/questions/6323603/ld-using-rpath-origin-inside-a-shared-library-recursive). As a consequence, it will not be able to resolve the symbols and report the missing symbols as errors, _even though the dynamic linker **would** be able to resolve these symbols_. We can tell the static linker not to report these errors by using `--unresolved-symbols=ignore-all` and all will be fine when we run the program ([(indeed, this is what the gold linker does)](https://sourceware.org/ml/binutils/2013-05/msg00038.html), but it makes the resulting library more difficult to use. 

Instead what we can do is make sure that the generated dynamic library has explicit top-level dependencies on these libraries. This means that the static linker knows where to find them, and when we have transitive dependencies on the same libraries the linker will only load them once, so we avoid needing to look at the `RPATH` of our dependencies. We can do this by passing `--no-as-needed` to the linker, so that it doesn't omit any libraries.

Note that on older ghc (7.6 and before) the Haskell libraries don't have an RPATH set at all, which makes it even more important that we make these top-level dependencies.

Finally, we have to explicitly link against `libffi` for the same reason. For newer ghc this _happens_ to be unnecessary on many systems because `libffi` is a library which is not specific to GHC, and when the static linker verifies that all symbols can be resolved it will find the `libffi` that is globally installed (completely independent from ghc). Of course, this may well be the _wrong_ version of `libffi`, but it's quite possible that symbol resolution happens to work. This is of course the wrong approach, which is why we link explicitly against `libffi` so that we will find the _right_ version of `libffi`.

**Comment by edsko** [comment]
So, the official status is now:

```
         7.4          7.6          7.8  7.10
Linux    OK           OK           OK   OK
Windows  Unknown      Unknown      OK   OK
OSX      Unsupported  Unsupported  OK   OK
```

7.4 and 7.6 on OSX have some path problems, somewhere in the package DB there are hardcoded references to `/usr/local/lib` even though I did not install anything there; however, we don't have the resources to address this right now so I suggest we open an issue and leave these as unsupported, unless somebody else wants to step up and fix it.

**Comment by 23Skidoo** [comment]
There should be a warning/error if the user tries to build a package containing platform libraries on Windows/OSX with GHC < 7.8. Aside from that, is there anything else to do here besides bikeshedding?

**Comment by edsko** [comment]
Not as far as I'm concerned. Although I'm not sure if @dcoutts has tested this yet for his original use-case (and it would be nice to get confirmation from @pawelp7 that this does in fact help him build his library (#2589).

**Comment by 23Skidoo** [comment]
Looking forward to merging this!

**Comment by pawelp7** [comment]
I wasn't able to make it working yet. I build cabal from your branch and used it to build my package. After failure I tried same with your tests (```Cabal\tests\PackageTests\PlatformLibs\my-platform-lib.cabal```):

```cabal install``` leads to:
```Warning: my-platform-lib.cabal: Ignoring unknown section type:
platform-library```

and than ```setup-Simple-Cabal-1.23.0.0-x86_64-windows-ghc-7.8.4.exe: No executables and no library found. Nothing to do.```

interestingly, removing argument from ```platform-library``` results in:
```cabal: my-platform-lib.cabal:9: 'platform-library' needs one argument (the
library's name)```
so it seems to nearly work :)

**Comment by edsko** [comment]
@pawelp7 If you get the "Unknown section type" error you are still using an old version of Cabal. My guess is that it's using an external setup method (possibly because you are using parallel builds? it's the default days) and hence your configure script gets built against the `Cabal` library that you have globally installed rather than the one that you built `cabal-install` against. To resolve this, try building with `-j1`, or if that still doesn't work, installing the `Cabal` library from my PR globally.

(This whole internal-versus-external setup thing is a major pain.)

**Comment by pawelp7** [comment]
@edsko Finally made it working. The problem was not in parallel build, but clearing cache and rebuilding cabal library and then ```cabal install``` made job. Thank you!

Cabal now generates the correct pair of libraries in ```dist/build/<package_name>```. However cabal still generates incorrect libraries ```HSdll.dll``` & ```HSdll.dll.a``` (missing symbols from .c file). Perhaps they should be removed or replaced by the correct ones, as their location is much more accessible?

**Comment by dcoutts** [comment]
@23Skidoo I think we're good to go.

**Comment by edsko** [comment]
Ok, I've rebased this PR, and merged it with my own fine-grained dependencies PR (#2514), as well as @nomeata 's PR to set RPATHs only when necessary (#2625) and @lukexi 's PR to avoid unnecessary recompilation of C files (#2601). I hope I didn't mess up in the process. I've also added the < 7.8 check for OSX, and renamed "platform libraries" to "foreign libraries". I've tested the rebased branch on OSX, Linux and Windows (just as a sanity check), and I think it's good to go. That said, I don't know how much scrutiny this PR has received yet; if not much, then it would be good in particular if my changes to `buildOrReplExe`, now called `gbuild`, would be verified (maybe by @dcoutts or @23Skidoo ?).

**Comment by 3noch** [comment]
How is this going? I'm getting DLLs on Windows but it's a bit hacky at this point and I have to ignore the fact that the build technically fails, but still produces a working DLL.

**Comment by edsko** [comment]
What do you mean by "and I have to ignore the fact that the build technically fails, but still produces a working DLL."?

**Comment by 3noch** [comment]
There is no way currently to give the DLL a name but I need to name the DLL differently for 32-bit and 64-bit builds. In order to do this, I specify a conditional `-o MyName.dll` flag which GHC complains about.

```
Building HsLibrary-0.1.0.0...
    Preprocessing library HsLibrary-0.1.0.0...
    Linking HsLibrary32.dll ...
    Creating library file: HsLibrary32.dll.a
    In-place registering HsLibrary-0.1.0.0...
    Installing library in
    C:\Code\windows-client\.stack-work\install\i386-windows\nightly-2015-07-06\7.10.1\lib\i386-windows-ghc-7.10.1\HsLib_KtkwOR4smXMBhy1EdkRbOf
    Setup.hs: 'C:\Users\Elliot
    Cameron\AppData\Local\Programs\stack\i386-windows\ghc-7.10.1\bin\ghc.exe'
    exited with an error:
    ghc.exe: can't apply -o to multiple source files
    Usage: For basic information, try the `--help' option.
```

**Comment by edsko** [comment]
Are you using this branch?

**Comment by edsko** [comment]
Ah, well, `foreign-library` takes an argument that corresponds to the lib name, but if you want a different lib name for 32-bit and 64-bit then you'll need two different `foreign-library` stanzas.

**Comment by 3noch** [comment]
If it merely put them in different `dist` folder's I'd be fine with that as well, but as it seems the current system allows zero configuration for DLLs. I can't even get them to show up in different builddirs!

**Comment by 3noch** [comment]
I'm not using your branch. I'm actually using `stack`.

**Comment by edsko** [comment]
Er.... then why are you commenting on this ticket? If you want support for platform libraries in `stack` you'll need to talk to the `stack` people :)

**Comment by 3noch** [comment]
Because `stack` still uses `Cabal` to pull this off as far as I know.

**Comment by 3noch** [comment]
I was using `cabal-install` prior to using `stack` and after switching I've gotten the exact same behavior as before. My `.cabal` file didn't change by one byte.

**Comment by edsko** [comment]
It's using a version of `Cabal` that has the changes in this branch applied? 

And as far as builddir is concerned, you cannot set this for individual artifacts. It seems the easiest solution to your problem is to define two separate `foreign-library` sections (provided that you are using this branch).

**Comment by 3noch** [comment]
I don't think `stack` is using a non-released version of `Cabal`, so unless I'm missing something, it doesn't have these features. I'd be fine with two stanzas even. It's a but of duplication but it's better than what I have now!

**Comment by 3noch** [comment]
In other words, if you release this, then `stack` can support these features too I hope.

**Comment by dcoutts** [comment]
@23Skidoo Any chance we could get this into 1.24 if we fix up the conflicts, or would you rather do this post 1.24?

**Comment by 23Skidoo** [comment]
@dcoutts Yes, I'd like to see this in 1.24. Would be nice if you could rebase it, then I won't have to do it myself :-)

**Comment by 23Skidoo** [comment]
@edsko @dcoutts Are you guys working on resolving conflicts or should I do it?

**Comment by 23Skidoo** [comment]
I probably won't have time to revive this before the 1.24 release, so postponing for after 1.24.

**Comment by 23Skidoo** [comment]
@hvr told me that GHC 8 final will likely be delayed until the end of March, so maybe we'll have time to get this in.

**Comment by blogle** [comment]
Just wanted to revive this thread to see if there was a time frame on when this might make it into release?

**Comment by ezyang** [comment]
I'll milestone it for 2.0, I don't see why we can't merge it for then.

**Comment by ezyang** [comment]
The conflicts get bigger and bigger. We should try to merge this sooner rather than later.

**Comment by crocket** [comment]
Why don't you call it `c-library`? It is explicit.

**Comment by ezyang** [comment]
Cleaning this up to merge for 2.0.

**Comment by ezyang** [comment]
@edsko, @dcoutts: If a package has foreign libraries but no libraries, should headers be installed?

**Comment by ezyang** [comment]
Subsumed by #4043.

**Comment by phadej** [comment]
So `standalone` option isn't supported?

**Comment by phadej** [comment]
To expand previous: Is adding `standalone` for Linux difficult task? Or why it isn't there in the first place?

**Pull request closed without merging**